File: services/agent-service/app/routers/skills.py
from fastapi import APIRouter, HTTPException, Response

from app.config import get_settings
from app.schemas import CreateSkillRequest, ListSkillsResponse, Skill, UpdateSkillRequest
from app.services.skill_service import (
    InvalidSkillNameError,
    SkillAlreadyExistsError,
    SkillNotFoundError,
    create_skill,
    delete_skill,
    list_skills,
    read_skill,
    update_skill,
)

router = APIRouter(prefix="/agent/skills", tags=["skills"])


@router.get("", response_model=ListSkillsResponse)
def get_skills() -> ListSkillsResponse:
    settings = get_settings()
    return ListSkillsResponse(skills=list_skills(settings.artifacts_root))


@router.get("/{name}", response_model=Skill)
def get_skill(name: str) -> Skill:
    settings = get_settings()
    try:
        skill = read_skill(settings.artifacts_root, name)
    except InvalidSkillNameError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    if skill is None:
        raise HTTPException(status_code=404, detail=f"Skill '{name}' not found")
    return Skill(**skill)


@router.post("", response_model=Skill, status_code=201)
def post_skill(request: CreateSkillRequest) -> Skill:
    settings = get_settings()
    try:
        skill = create_skill(settings.artifacts_root, request.name, request.description, request.content)
    except InvalidSkillNameError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except SkillAlreadyExistsError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    return Skill(**skill)


@router.put("/{name}", response_model=Skill)
def put_skill(name: str, request: UpdateSkillRequest) -> Skill:
    settings = get_settings()
    try:
        skill = update_skill(settings.artifacts_root, name, request.description, request.content)
    except InvalidSkillNameError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except SkillNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return Skill(**skill)


@router.delete("/{name}", status_code=204, response_class=Response)
def remove_skill(name: str) -> Response:
    settings = get_settings()
    try:
        delete_skill(settings.artifacts_root, name)
    except InvalidSkillNameError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except SkillNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return Response(status_code=204)

File: services/artifact-agent-service/src/routes/skills.ts
import type { FastifyInstance } from 'fastify';
import type { AppConfig } from '../config.js';
import { parseCreateSkillRequest, parseUpdateSkillRequest, ValidationError } from '../schemas.js';
import {
  createSkill,
  deleteSkill,
  InvalidSkillNameError,
  listSkills,
  readSkill,
  SkillAlreadyExistsError,
  SkillNotFoundError,
  updateSkill,
} from '../services/skill-service.js';

interface SkillNameParams {
  name: string;
}

export function registerSkillsRoutes(fastify: FastifyInstance, config: AppConfig): void {
  fastify.get('/agent/skills', async () => ({ skills: listSkills(config.artifactsRoot) }));

  fastify.get<{ Params: SkillNameParams }>('/agent/skills/:name', async (request, reply) => {
    try {
      const skill = readSkill(config.artifactsRoot, request.params.name);
      if (!skill) {
        return reply.code(404).send({ detail: `Skill '${request.params.name}' not found` });
      }
      return skill;
    } catch (err) {
      if (err instanceof InvalidSkillNameError) {
        return reply.code(400).send({ detail: err.message });
      }
      throw err;
    }
  });

  fastify.post('/agent/skills', async (request, reply) => {
    let parsed;
    try {
      parsed = parseCreateSkillRequest(request.body);
    } catch (err) {
      if (err instanceof ValidationError) {
        return reply.code(422).send({ detail: err.message });
      }
      throw err;
    }

    try {
      const skill = createSkill(config.artifactsRoot, parsed.name, parsed.description, parsed.content);
      return reply.code(201).send(skill);
    } catch (err) {
      if (err instanceof InvalidSkillNameError) {
        return reply.code(400).send({ detail: err.message });
      }
      if (err instanceof SkillAlreadyExistsError) {
        return reply.code(409).send({ detail: err.message });
      }
      throw err;
    }
  });

  fastify.put<{ Params: SkillNameParams }>('/agent/skills/:name', async (request, reply) => {
    let parsed;
    try {
      parsed = parseUpdateSkillRequest(request.body);
    } catch (err) {
      if (err instanceof ValidationError) {
        return reply.code(422).send({ detail: err.message });
      }
      throw err;
    }

    try {
      return updateSkill(config.artifactsRoot, request.params.name, parsed.description, parsed.content);
    } catch (err) {
      if (err instanceof InvalidSkillNameError) {
        return reply.code(400).send({ detail: err.message });
      }
      if (err instanceof SkillNotFoundError) {
        return reply.code(404).send({ detail: err.message });
      }
      throw err;
    }
  });

  fastify.delete<{ Params: SkillNameParams }>('/agent/skills/:name', async (request, reply) => {
    try {
      deleteSkill(config.artifactsRoot, request.params.name);
      return reply.code(204).send();
    } catch (err) {
      if (err instanceof InvalidSkillNameError) {
        return reply.code(400).send({ detail: err.message });
      }
      if (err instanceof SkillNotFoundError) {
        return reply.code(404).send({ detail: err.message });
      }
      throw err;
    }
  });
}
